File: c++/Write_bit.cpp
#include "Write_bit.h"

Write_bit::Write_bit(const std::string &fileName){
    write_stream.open(fileName);
    buffer=0;
    bits_in_buffer=0;
}

void Write_bit::operator<<(const bool &bit){
    if (bit)
        buffer |= (1<<(CHAR_BIT-(bits_in_buffer+1)));
    bits_in_buffer++;
    //std::cout << bit << "\t" << bits_in_buffer_in << std::endl;
    if (bits_in_buffer == CHAR_BIT){
        //std::cout << buffer_in << std::endl;
        write_stream << buffer;
        bits_in_buffer = 0;
        buffer = 0;
    }
    return;
}

void Write_bit::close(){
    for(int i=bits_in_buffer; i<CHAR_BIT;i++){
        this->operator<<(0);
    }
    write_stream<<"\x0a";
    write_stream.close();
    return;
}

File: python/Write_bit.py
class Write_bit:
    def __init__(self,namefile):
        self.namefile = namefile
        self.file = open(namefile,"wb")
        self.buffer=0;
        self.bits_in_buffer=0;
        self.CHAR_BIT = 8

    def write(self,bit):
        if(bit):
            self.buffer |= (1<<(self.CHAR_BIT-(self.bits_in_buffer+1)))
        self.bits_in_buffer+=1
        if(self.bits_in_buffer == self.CHAR_BIT):
            self.file.write(chr(self.buffer))
            self.buffer=0
            self.bits_in_buffer=0
        return

    def close(self):
        for i in range(self.bits_in_buffer, self.CHAR_BIT):
            self.write(0)
        #self.file.write("\x00")
        self.file.close();
        return



